R/recursive_unit_tests.R

```diff
@@ -0,0 +1,192 @@
+      private$cat_cr()
+
+      cli::cat_line(
+        private$colourise(cli::symbol$tick, "success"), " | ",
+        private$colourise("F", "failure"), " ",
+        private$colourise("W", "warning"), " ",
+        private$colourise("S", "skip"), " ",
+        private$colourise(" OK", "success"),
+        " | ", "Test Directory"
+      )
+
+      summary_results <- as.data.frame(test_results)
+
+      sapply(row.names(summary_results), function(summary_row_name) {
+        path_test_result <- summary_results[summary_row_name, ]
+
+        if (path_test_result$failed > 0) {
+          status <- cli::col_red(cli::symbol$cross)
+        } else {
+          status <- cli::col_green(cli::symbol$tick)
+        }
+
+        message <- paste0(
+          status, " | ",
+          private$col_format(path_test_result$failed, "fail"), " ",
+          private$col_format(path_test_result$warning, "warn"), " ",
+          private$col_format(path_test_result$skipped, "skip"), " ",
+          sprintf("%3d", path_test_result$passed),
+          " | ", summary_row_name
+        )
+
+        cli::cat_line(message)
+      })
+
+      private$cat_cr()
+    },
+    test_recursive = function(...) {
+      test_results <- private$run_recursive_test_dir(...)
+
+      private$show_summary(test_results)
+
+      private$show_final_line(test_results)
+    },
+    cat_cr = function() {
+      if (cli::is_dynamic_tty()) {
+        cli::cat_line("\r")
+      } else {
+        cli::cat_line("\n")
+      }
+    },
+    col_format = function(n, type) {
+      if (n == 0) {
+        " "
+      } else {
+        private$colourise(n, type)
+      }
+    },
+    colourise = function(text, as = c("success", "skip", "warning", "failure", "error")) {
+      if (private$has_colour()) {
+        unclass(cli::make_ansi_style(private$testthat_style(as))(text))
+      } else {
+        text
+      }
+    },
+    has_colour = function() {
+      isTRUE(getOption("testthat.use_colours", TRUE)) &&
+        cli::num_ansi_colors() > 1
+    },
+    summary_line = function(n_fail, n_warn, n_skip, n_pass) {
+      colourise_if <- function(text, colour, cond) {
+        if (cond) private$colourise(text, colour) else text
+      }
+
+      # Ordered from most important to least important
+      paste0(
+        "[ ",
+        colourise_if("FAIL", "failure", n_fail > 0), " ", n_fail, " | ",
+        colourise_if("WARN", "warn", n_warn > 0),    " ", n_warn, " | ",
+        colourise_if("SKIP", "skip", n_skip > 0),    " ", n_skip, " | ",
+        colourise_if("PASS", "success", n_fail == 0), " ", n_pass,
+        " ]"
+      )
+    },
+    testthat_style = function(type = c("success", "skip", "warning", "failure", "error")) {
+      type <- match.arg(type)
+
+      c(
+        success = "green",
+        skip = "blue",
+        warning = "magenta",
+        failure = "orange",
+        error = "orange"
+      )[[type]]
+    }
+  )
+)
+
+r_cmd_check_fix <- function() {
+  testthat::test_check()
+}
```

Comment:
Without this, I was getting an R CMD CHECK note `Namespace in Imports field not imported from: 'testthat'`